windows/src/engine/mcompile/mc_unicode.cpp

```diff
@@ -276,7 +276,7 @@ BOOL ConvertKeyboardToUnicode(LPKEYBOARD kbd) {
   LPKEY kp;
   DWORD i, j;
 
-  if(kbd->StartGroup[BEGIN_UNICODE] >= 0 && kbd->StartGroup[BEGIN_UNICODE] < kbd->cxGroupArray) {
```

Comment:
So... I assume this means that `kbd->StartGroup[BEGIN_UNICODE]` is set to -1 or something like that when it is not available?  I'd normally expect a `0` for a null state, though I understand if that's not the case here for legacy reasons.

Reply:
`kbd->StartGroup[BEGIN_UNICODE]` are unsigned long, so set to `0xFFFFFFFFL` (`-1`) on line 461 of Compiler.cpp. The `>= 0` test is redundant because of unsigned. This is an index into an array, so `-1` is equivalent to null state.

Reply:
So the remaining check still covers the `null` case?  It's not immediately clear that `kbd->StartGroup[BEGIN_UNICODE] == 0xFFFFFFFFL` can only be the case if `kbd->cxGroupArray` also holds that value, which is the one way I can see the "`null`" case still being handled here.

Reply:
This is testing that `kbd->StartGroup` is in a valid range (`0 .. kbd->cxGroupArray-1`), and `0xFFFFFFFFL` is always going to be outside the valid range.